content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md

```diff
@@ -0,0 +1,383 @@
+
+2. Get a list of component types. Call `AzToolsFramework::EditorComponentAPIBus::BroadcastResult(...)` and dispatch `AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeIdsByService`, which takes a list of services to include (`providedServices`) and services to exclude (an empty list) and stores the component types to `typeIds`.
+
+```cpp
+AzToolsFramework::EntityCompositionRequests::ComponentServicesList providedServices;
+providedServices.push_back(AZ_CRC_CE("ShapeService"));
+
+AZ::ComponentTypeList typeIds;
+AzToolsFramework::EditorComponentAPIBus::BroadcastResult(
+    typeIds,
+    &AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeIdsByService,
+    providedServices,
+    AzToolsFramework::EntityCompositionRequests::ComponentServicesList()
+);
+```
+
+
+Next, get the names of the components to add to the buttons. 
+
+1. Get a list of component names. Call `AzToolsFramework::EditorComponentAPIBus::BroadcastResult(...)` and dispatch `AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeNames`, which takes the list of component types (`typeIds`) and stores the associated of component names to `componentNames`.
+
+```cpp
+AZStd::vector<AZStd::string> componentNames;
+AzToolsFramework::EditorComponentAPIBus::BroadcastResult(
+    componentNames,
+    &AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeNames,
+    typeIds
+);
+```
+
+{{ note }}
+You can query a list of services that components provide by calling the component's `GetProvidedServices()` method.
+{{ /note }}
+
+### Add buttons
+
+Next, you will loop through the list of components, create buttons, and add them to the `gridLayout`. Refer to the following snippet in `ShapeEXampleQWidget.cpp`.
+
+For each component:
+
+1. Store the component's name and type from the list of component names and types that you queried earlier.
+
+2. Create a button using the QButton object called `shapeButton`. 
+   - Pass in `name` to add the component's name onto the button. 
+   - Set the button minimum height to `40` by using `setMinimumHeight(...)`. 
+
+3. Later, you will connect `shapeButton` to a signal listener. 
+
+4. Add the button to `gridLayout`, which is split into three columns. 
+
+```cpp
+const int maxColumnCount = 3;
+for (int i = 0; i < componentNames.size(); ++i)
+{
+    AZStd::string name = componentNames[i];
+    AZ::TypeId typeId = typeIds[i];
+
+    QPushButton* shapeButton = new QPushButton(QString::fromUtf8(name.c_str()), this);
+    shapeButton->setMinimumHeight(40);
+
+    // Connect 
+
+    int row = i / maxColumnCount;
+    int column = i % maxColumnCount;
+    gridLayout->addWidget(shapeButton, row, column);
+}
+
+```
+
+### Add a signal listener with a lambda handler
+
+Next, you will create a signal listener such that when the user clicks a button, an entity with the corresponding Shape component will spawn in your scene. 
```

Comment:
```suggestion
Next, you will create a signal listener such that when the user clicks a button, an entity will be created in your scene and the corresponding Shape component will be added to that entity.
```

I rephrased this slightly to not use the word "spawn", since that has a meaning specific to prefabs.